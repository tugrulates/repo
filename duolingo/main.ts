import { Command } from "@cliffy/command";
import { Input, Secret } from "@cliffy/prompt";
import { Table } from "@cliffy/table";
import { pool } from "@roka/async/pool";
import { type Config, config } from "@roka/cli/config";
import { version } from "@roka/forge/version";
import { DuolingoClient } from "./client.ts";
import { LEAGUES } from "./data.ts";
import {
  engageWithCard,
  followLeagueUsers,
  getEmoji,
  getLeagueUserEmoji,
} from "./interaction.ts";
import type { FeedCard, League } from "./types.ts";

type DuolingoConfig = { username: string; token: string };

/** Duolingo client built from common CLI options. */
async function getClient(cfg: Config<DuolingoConfig>): Promise<DuolingoClient> {
  let { username, token } = await cfg.get();
  if (!username) username = await Input.prompt("Username");
  if (!token) token = await Secret.prompt("Token");
  if (!username || !token) throw new Error("Username and token are required.");
  return new DuolingoClient(username, token);
}

/**
 * Fetches and organizes follow information.
 *
 * @returns Users who are followed, users who follow, and their difference sets.
 */
async function getFollows(client: DuolingoClient) {
  const [following, followers] = await Promise.all([
    client.getFollowing(),
    client.getFollowers(),
  ]);
  return {
    following,
    followers,
    dontFollowBack: following.filter(({ userId }) =>
      !followers.some((user) => user.userId === userId)
    ),
    notFollowingBack: followers.filter(({ userId }) =>
      !following.some((user) => user.userId === userId)
    ),
  };
}

/**
 * Outputs the league to the console.
 *
 * @param league League to output.
 */
async function outputLeague(client: DuolingoClient, league: League) {
  const following = await client.getFollowing();
  const tier = LEAGUES[league.tier];
  new Table()
    .header([tier.emoji, tier.name])
    .body(
      league.rankings.map((user, index) => [
        `${index + 1}.`,
        `${user.display_name} ${getLeagueUserEmoji(user)}`,
        following.find((f) => f.userId === user.user_id) ? "üë§" : "",
        `${user.score.toString()} XP`,
      ]),
    )
    .columns([{ align: "right" }, {}, {}, { align: "right" }])
    .render();
}

/**
 * Returns the display summary of the card.
 *
 * @param card Card to get the summary for.
 * @returns Display summary of the card.
 */
function getSummary(card: FeedCard): string {
  return card.header
    ? card.header
      .replace(/[\u200E-\u200F]/g, "")
      .replace(/<[^>]+>/g, "")
    : `${card.displayName} ${card.body.toLowerCase()}`;
}

function getFeedCommand(cfg: Config<DuolingoConfig>) {
  return new Command()
    .description("Prints and interacts with the feed.")
    .example("duolingo feed", "Prints the feed.")
    .example("duolingo feed --engage", "Engages with the feed.")
    .example("duolingo feed --json | jq", "Query JSON over the feed.")
    .option("--engage", "Engage with the feed events.")
    .option("--json", "Output the feed as JSON.")
    .action(async ({ engage, json }) => {
      const client = await getClient(cfg);
      const followers = await client.getFollowers();
      const cards = await client.getFeedCards();
      if (json) console.log(JSON.stringify(cards, undefined, 2));
      await pool(
        cards,
        async (card) => {
          if (!engage || await engageWithCard(client, followers, card)) {
            if (!json) console.log(`${getEmoji(card)} ${getSummary(card)}`);
          }
        },
      );
    });
}

function getFollowsCommand(config: Config<DuolingoConfig>) {
  return new Command()
    .description("Prints and manages follower information on Duolingo.")
    .example("duolingo follows", "Prints follow counts.")
    .example("duolingo follows --follows", "Follow users who follow.")
    .example("duolingo follows --unfollow", "Unfollow users who dont' follow.")
    .example("duolingo follows --follow --unfollow", "Matches both lists.")
    .example("duolingo follows --json", "Outputs JSON of follower information.")
    .example(
      "duolingo follows --json | jq",
      "Query JSON for follower information.",
    )
    .example(
      "duolingo follows --json | jq '.dontFollowBack[].username'",
      "List users who are followed but don't follow back.",
    )
    .example(
      "duolingo follows --json | jq '.notFollowingBack[].username'",
      "List users who follow but are not followed back.",
    )
    .option("--follow", "Follow users who follow.")
    .option("--unfollow", "Unfollow users who don't follow.")
    .option("--json", "Output the follower information as JSON.")
    .action(async ({ follow, unfollow, json }) => {
      const client = await getClient(config);
      let result = await getFollows(client);

      if (follow || unfollow) {
        if (follow) {
          await pool(
            result.notFollowingBack,
            async (user) => {
              if (user.canFollow) {
                await client.followUser(user.userId);
                if (!json) console.log(`‚úÖ Followed ${user.username}.`);
              }
            },
          );
        }
        if (unfollow) {
          await pool(
            result.dontFollowBack,
            async (user) => {
              await client.unfollowUser(user.userId);
              if (!json) console.log(`‚ùå Unfollowed ${user.username}.`);
            },
          );
        }
        result = await getFollows(client);
      }

      if (json) console.log(JSON.stringify(result, undefined, 2));
      else {
        console.log(`üë§ Following ${result.following.length} people.`);
        console.log(`üë§ Followed by ${result.followers.length} people.`);
      }
    });
}

function getLeagueCommand(config: Config<DuolingoConfig>) {
  return new Command()
    .description("Prints and interacts with the current Duolingo league.")
    .example("duolingo league", "Prints the league.")
    .example("duolingo league --follow", "Follows users in the league.")
    .example("duolingo league --json | jq", "Query JSON over the league.")
    .option("--follow", "Follow users in the league.")
    .option("--json", "Output the league as JSON.")
    .action(async ({ follow, json }) => {
      const client = await getClient(config);
      const league = await client.getLeague();
      if (league) {
        if (follow) await followLeagueUsers(client, league.rankings);
        if (json) console.log(JSON.stringify(league, undefined, 2));
        else await outputLeague(client, league);
      } else {
        if (json) console.log("{}");
        else console.log("üèÜ The league has not started yet.");
      }
    });
}

async function getCommand(cfg: Config<DuolingoConfig>) {
  const { username, token } = await cfg.get();
  const command = new Command()
    .name("duolingo")
    .description("Interact with Duolingo.")
    .usage("--username <username> --token <token> <command> [options]")
    .version(await version({ target: true }))
    .example("duolingo --username <username> --token <token>", "Configure.")
    .example("duolingo --clear", "Clear the cached configuration.")
    .option("--clear", "Clear the cached configuration.", {
      standalone: true,
      action: () => cfg.clear(),
    })
    .globalEnv(
      "DUOLINGO_USERNAME=<username:string>",
      "Username.",
      { prefix: "DUOLINGO_" },
    )
    .globalOption(
      "--username <username:string>",
      "Username.",
      username ? { default: username } : {},
    )
    .globalEnv(
      "DUOLINGO_TOKEN=<token:string>",
      "JWT token.",
      { prefix: "DUOLINGO_" },
    )
    .globalOption(
      "--token <token:string>",
      "JWT token.",
      token ? { default: token } : {},
    )
    .help({ colors: Deno.stdout.isTerminal() })
    .noExit()
    .globalAction((options) => cfg.set(options))
    .command("feed", getFeedCommand(cfg))
    .command("follows", getFollowsCommand(cfg))
    .command("league", getLeagueCommand(cfg));
  return command;
}

/** CLI entrypoint. */
export async function main(args: string[], options?: { path?: string }) {
  using cfg = config<DuolingoConfig>(options);
  const command = await getCommand(cfg);
  await command.parse(args);
}
