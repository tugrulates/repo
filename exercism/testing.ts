import { assertExists } from "@std/assert/exists";
import { pick } from "@std/collections";
import { toPascalCase } from "@std/text";
import type {
  Client,
  CompletionData,
  ConceptData,
  ConceptProgressionData,
  ExerciseData,
  FileContent,
  FilesData,
  IterationData,
  ReputationData,
  ReputationMeta,
  SolutionData,
  SubmissionData,
  TaskData,
  TestResultData,
  TestRunData,
  TrackData,
  UserData,
} from "./client.ts";

/** Options for the {@link fakeClient} function. */
export interface FakeClientOptions {
  /** Whether to validate or invalidate the authentication token. */
  validate?: boolean;
  /** The user to return from the API. */
  user?: UserData;
  /** The reputation to return from the API. */
  reputation?: ReputationMeta;
  /** The tracks to return from the API. */
  tracks?: TrackData[];
  /** The exercises to return from the API. */
  exercises?: ExerciseData[];
  /** The solutions to return from the API. */
  solutions?: SolutionData[];
  /** The iterations to return from the API. */
  iterations?: IterationData[];
  /** The files to return from the API. */
  files?: FilesData;
  /** The file contents to return from the API. */
  fileContents?: FileContent[];
}

/** Creates a fake Exercism client for testing purposes. */
export function fakeClient(options?: FakeClientOptions): Client {
  let {
    validate,
    user,
    reputation,
    tracks,
    exercises,
    solutions,
    iterations,
    files,
    fileContents,
  } = options ?? {};
  const client: Client = {
    token: {
      async validate() {
        assertExists(validate, "Validate not provided");
        await Promise.resolve();
        return validate;
      },
    },
    user: {
      async get() {
        assertExists(user, "User not provided");
        await Promise.resolve();
        return { user };
      },
      async reputation() {
        assertExists(reputation, "Reputation not provided");
        await Promise.resolve();
        return {
          results: [
            testReputationData({ uuid: "test-reputation-uuid-1" }),
            testReputationData({ uuid: "test-reputation-uuid-2" }),
          ],
          meta: reputation,
        };
      },
    },
    tracks: {
      async list() {
        assertExists(tracks, "Tracks not provided");
        await Promise.resolve();
        return { tracks };
      },
    },
    track: (_) => ({
      exercises: {
        async list() {
          assertExists(exercises, "Exercises not provided");
          await Promise.resolve();
          return { exercises, solutions: solutions ?? [] };
        },
      },
      exercise(slug) {
        return {
          async start() {
            await Promise.resolve();
            const solution = testSolutionData({
              exercise: { slug },
              status: "started",
              num_iterations: 0,
            });
            if (!solutions) solutions = [];
            solutions.push(solution);
            return { solution };
          },
        };
      },
    }),
    solution: (uuid) => ({
      async get() {
        const solution = solutions?.find((solution) => solution.uuid === uuid);
        assertExists(solution, "Solution not provided");
        await Promise.resolve();
        return { solution };
      },
      async complete() {
        const solution = solutions?.find((solution) => solution.uuid === uuid);
        assertExists(solution, "Solution not provided");
        if (solution.status === "iterated") solution.status = "completed";
        await Promise.resolve();
        return testCompletionData();
      },
      async publish() {
        const solution = solutions?.find((solution) => solution.uuid === uuid);
        assertExists(solution, "Solution not provided");
        solution.status = "published";
        await Promise.resolve();
        return { solution };
      },
      async sync() {
        await Promise.resolve();
        return { solution: testSolutionData({ uuid, status: "iterated" }) };
      },
      submissions: {
        async create() {
          await Promise.resolve();
          return { submission: testSubmissionData({ uuid }) };
        },
        async testRun() {
          await Promise.resolve();
          return { test_run: testTestRunData({ uuid }) };
        },
      },
      iterations: {
        async create() {
          const solution = solutions?.find((solution) =>
            solution.uuid === uuid
          );
          assertExists(solution, "Solution not provided");
          if (solution?.status === "started") solution.status = "iterated";
          await Promise.resolve();
          return { iteration: testIterationData({ uuid }) };
        },
        async latest() {
          const iteration = iterations?.find((iteration) =>
            iteration.uuid === uuid
          );
          assertExists(iteration, "Iteration not provided");
          await Promise.resolve();
          return { iteration };
        },
        async publish() {
          const iteration = iterations?.find((iteration) =>
            iteration.uuid === uuid
          );
          assertExists(iteration, "Iteration not provided");
          iteration.is_published = true;
          await Promise.resolve();
          return { iteration };
        },
      },
      files: {
        async list() {
          assertExists(files, "Files not provided");
          await Promise.resolve();
          return { files };
        },
        async submission() {
          assertExists(fileContents, "File contents not provided");
          await Promise.resolve();
          return { files: fileContents };
        },
        async iteration() {
          assertExists(fileContents, "File contents not provided");
          await Promise.resolve();
          return { files: fileContents };
        },
      },
    }),
  };
  return client;
}

/** Creates a dummy {@link UserData} object for testing. */
export function testUserData(data?: Partial<UserData>): UserData {
  return {
    handle: "test-user-handle",
    insiders_status: "ineligible",
    ...data,
  };
}

/** Creates a dummy {@link ReputationData} object for testing. */
export function testReputationData(
  data?: Partial<ReputationData>,
): ReputationData {
  return {
    uuid: "test-reputation-uuid",
    value: 1,
    text: "test-reputation-text",
    icon_url: "test/reputation.svg",
    internal_url: "test/reputation",
    external_url: "test/user/reputation",
    created_at: "2000-01-01T00:00:00Z",
    track: pick(testTrackData(), ["title", "icon_url"]),
    is_seen: false,
    links: {
      mark_as_seen: "test/reputation/mark_as_seen",
    },
    ...data,
  };
}

/** Creates a dummy {@link ReputationMeta} object for testing. */
export function testReputationMeta(
  data?: Partial<ReputationMeta>,
): ReputationMeta {
  return {
    current_page: 1,
    total_count: 123,
    total_pages: 12,
    total_reputation: 42,
    unseen_total: 1,
    links: {
      tokens: "test/reputation/tokens",
      mark_all_as_seen: "test/reputation/mark_all_as_seen",
    },
    ...data,
  };
}

/** Creates a dummy {@link TrackData} object for testing. */
export function testTrackData(data?: Partial<TrackData>): TrackData {
  const { slug = "test-track-slug" } = data ?? {};
  return {
    slug,
    title: toPascalCase(slug),
    course: true,
    num_concepts: 12,
    num_exercises: 42,
    web_url: `test/${slug}`,
    icon_url: `test/${slug}.svg`,
    tags: ["test-tag-1", "test-tag-2"],
    last_touched_at: "2000-01-01T00:00:00Z",
    is_new: false,
    links: {
      self: `test/track/${slug}`,
      exercises: `test/track/${slug}/exercises`,
      concepts: `test/track/${slug}/concepts`,
    },
    ...data,
  };
}

/** Creates a dummy {@link ExerciseData} object for testing. */
export function testExerciseData(data?: Partial<ExerciseData>): ExerciseData {
  const { slug = "test-exercise-slug" } = data ?? {};
  return {
    slug,
    type: "practice",
    title: toPascalCase(slug),
    icon_url: `test/${slug}.svg`,
    difficulty: "easy",
    blurb: "test-exercise-blurb",
    is_external: true,
    is_unlocked: false,
    is_recommended: false,
    links: {
      self: `test/exercise/${slug}`,
    },
    ...data,
  };
}

/** Creates a dummy {@link ConceptData} object for testing. */
export function testConceptData(data?: Partial<ConceptData>): ConceptData {
  const { slug = "test-concept-slug" } = data ?? {};
  return {
    slug,
    name: toPascalCase(slug),
    links: {
      self: `test/concept/${slug}`,
    },
    ...data,
  };
}

/** Creates a dummy {@link TaskData} object for testing. */
export function testTaskData(data?: Partial<TaskData>): TaskData {
  return {
    id: 1,
    title: "test-task-title",
    ...data,
  };
}

/** Creates a dummy {@link SolutionData} object for testing. */
export function testSolutionData(data?: Partial<SolutionData>): SolutionData {
  const { uuid = "test-solution-uuid", status = "started" } = data ?? {};
  const published = status === "published";
  const completed = status === "completed" || published;
  const iterated = status === "iterated" || completed;
  return {
    uuid,
    private_url: `test/solution/${uuid}`,
    public_url: `test/solution/${uuid}/public`,
    status,
    mentoring_status: "none",
    published_iteration_head_tests_status: published ? "passed" : "not_queued",
    has_notifications: false,
    num_views: 0,
    num_stars: 0,
    num_comments: 0,
    num_iterations: iterated ? 1 : 0,
    num_loc: 123,
    is_out_of_date: false,
    published_at: published ? "2000-01-01T00:00:00Z" : null,
    completed_at: completed ? "2000-01-01T00:00:00Z" : null,
    updated_at: "2000-01-01T00:00:00Z",
    last_iterated_at: iterated ? "2000-01-01T00:00:00Z" : null,
    track: pick(testTrackData(), ["slug", "title", "icon_url"]),
    exercise: pick(testExerciseData(), ["slug", "title", "icon_url"]),
    ...data,
  };
}

/** Creates a dummy {@link SubmissionData} object for testing. */
export function testSubmissionData(
  data?: Partial<SubmissionData>,
): SubmissionData {
  const { uuid = "test-submission-uuid" } = data ?? {};
  return {
    uuid,
    tests_status: "passed",
    links: {
      cancel: `test/submission/${uuid}/cancel`,
      submit: `test/submission/${uuid}/submit`,
      test_run: `test/submission/${uuid}/test_run`,
      ai_help: `test/submission/${uuid}/ai_help`,
      initial_files: `test/submission/${uuid}/initial_files`,
      last_iteration_files: `test/submission/${uuid}/last_iteration_files`,
    },
    ...data,
  };
}

/** Creates a dummy {@link TestRunData} object for testing. */
export function testTestRunData(data?: Partial<TestRunData>): TestRunData {
  const { uuid = "test-test-run-uuid" } = data ?? {};
  return {
    uuid,
    submission_uuid: "test-submission-uuid",
    version: 1,
    status: "pass",
    message: null,
    message_html: null,
    output: null,
    output_html: null,
    tasks: [testTaskData({ id: 1 }), testTaskData({ id: 2 })],
    highlightjs_language: "test-highlightjs-language",
    tests: [
      testTestResultData({ task_id: 1 }),
      testTestResultData({ task_id: 2 }),
    ],
    links: {
      self: `test/test_run/${uuid}`,
    },
    ...data,
  };
}

/** Creates a dummy {@link TestResultData} object for testing. */
export function testTestResultData(
  data?: Partial<TestResultData>,
): TestResultData {
  return {
    name: "test-test-result-name",
    status: "pass",
    test_code: "test-test-result-test-code",
    message: "test-test-result-message",
    message_html: "test-test-result-message-html",
    expected: "test-test-result-expected",
    output: "test-test-result-output",
    output_html: "test-test-result-output-html",
    task_id: null,
    ...data,
  };
}

/** Creates a dummy {@link CompletionData} object for testing. */
export function testCompletionData(
  data?: Partial<CompletionData>,
): CompletionData {
  return {
    track: testTrackData(),
    exercise: testExerciseData(),
    unlocked_exercises: [testExerciseData({ slug: "unlocked" })],
    unlocked_concepts: [testConceptData({ slug: "unlocked" })],
    concept_progressions: [testConceptProgressionData({ slug: "progressed" })],
    ...data,
  };
}

/** Creates a dummy {@link ConceptProgressionData} object for testing. */
export function testConceptProgressionData(
  data?: Partial<ConceptProgressionData>,
): ConceptProgressionData {
  const { slug = "test-concept-slug" } = data ?? {};
  return {
    slug,
    name: toPascalCase(slug),
    from: 0,
    to: 1,
    total: 2,
    links: {
      self: `test/concept/${slug}`,
    },
    ...data,
  };
}

/** Creates a dummy {@link IterationData} object for testing. */
export function testIterationData(
  data?: Partial<IterationData>,
): IterationData {
  const { uuid = "test-solution-uuid" } = data ?? {};
  return {
    uuid,
    submission_uuid: "test-submission-uuid",
    idx: 1,
    status: null,
    num_essential_automated_comments: 0,
    num_actionable_automated_comments: 0,
    num_non_actionable_automated_comments: 0,
    num_celebratory_automated_comments: 0,
    submission_method: "api",
    created_at: "2000-01-01T00:00:00Z",
    tests_status: "passed",
    is_published: false,
    is_latest: true,
    links: {
      self: `test/iteration/${uuid}/iteration`,
      automated_feedback: `test/iteration/${uuid}/automated_feedback`,
      delete: `test/iteration/${uuid}/delete`,
      solution: `test/iteration/${uuid}`,
      test_run: `test/iteration/${uuid}/test_run`,
      files: `test/iteration/${uuid}/files`,
    },
    ...data,
  };
}

/** Creates a dummy {@link FilesData} object for testing. */
export function testFilesData(data?: Partial<FilesData>): FilesData {
  return {
    solution: ["test-solution-file.ts"],
    test: ["test-test-file.ts"],
    example: [".meta/test-example-file.ts"],
    ...data,
  };
}

/** Creates a dummy {@link FileContent} object for testing. */
export function testFileContent(data?: Partial<FileContent>): FileContent {
  return {
    filename: "test-filename.ts",
    content: "test-content",
    type: "solution",
    ...data,
  };
}
