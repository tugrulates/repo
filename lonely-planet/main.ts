// deno-lint-ignore-file no-console
import { Command } from "@cliffy/command";
import { Table } from "@cliffy/table";
import { displayVersion } from "@roka/package/version";
import { LonelyPlanetClient } from "./client.ts";
import { EMOJIS } from "./data.ts";
import type { Attraction, Destination } from "./types.ts";

/** Breadcrumb text from a location. */
function breadcrumb(document: Destination | Attraction) {
  return `[ ${
    document.breadcrumb.map((breadcrumb) => breadcrumb.title).join(" > ")
  } ]`;
}

async function getCommand() {
  return new Command()
    .name("lonely-planet")
    .description("Explores data from Lonely Planet.")
    .version(await displayVersion())
    .example("lonely-planet big sur", "Search destinations for 'big sur'.")
    .example("lonely-planet --attractions amsterdam", "Search attractions.")
    .example("lonely-planet --stories amsterdam", "Search stories.")
    .example("lonely-planet --destinations --attractions --stories", "All.")
    .example("lonely-planet --json | jq", "Stream destinations as json.")
    .arguments("[keywords...:string]")
    .option("--destinations", "Include destinations in the results.")
    .option("--attractions", "Include attractions in the results.")
    .option("--stories", "Include stories in the results.")
    .option("--json", "Output the search results as concatenated JSON.")
    .action(
      async ({ destinations, attractions, stories, json }, ...keywords) => {
        if (!destinations && !attractions && !stories) destinations = true;
        const client = new LonelyPlanetClient();
        const rows: string[][] = [];
        if (destinations) {
          for await (const doc of client.searchDestinations(keywords)) {
            if (json) console.log(JSON.stringify(doc));
            else rows.push([EMOJIS[doc.type], doc.title, breadcrumb(doc)]);
          }
        }
        if (attractions) {
          for await (const doc of client.searchAttractions(keywords)) {
            if (json) console.log(JSON.stringify(doc));
            else rows.push([EMOJIS[doc.type], doc.title, breadcrumb(doc)]);
          }
        }
        if (stories) {
          for await (const doc of client.searchStories(keywords)) {
            if (json) console.log(JSON.stringify(doc));
            else rows.push([EMOJIS[doc.type], doc.title]);
          }
        }
        Table.from(rows).render();
      },
    );
}

/** CLI entrypoint. */
export async function main(args: string[]) {
  const command = await getCommand();
  await command.parse(args);
}
