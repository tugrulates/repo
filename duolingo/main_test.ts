// deno-lint-ignore-file no-console
import { mockFetch } from "@roka/http/testing";
import { fakeConsole } from "@roka/testing/fake";
import type { Mock } from "@roka/testing/mock";
import { assertSnapshot } from "@std/testing/snapshot";
import { main } from "./main.ts";

// Use ENV variables for recording, but fake credentials for replay.
const CONFIG = ["--username", "TugrulAtes", "--token", "TOKEN"];

function config(mock: Mock<typeof fetch>) {
  if (mock.mode === "replay") {
    return CONFIG;
  }
  return [];
}

Deno.test("duolingo feed", async (t) => {
  using console = fakeConsole();
  using fetch = mockFetch(t);
  await main([...config(fetch), "feed"], { path: ":memory:" });
  await assertSnapshot(t, console.calls);
});

Deno.test("duolingo feed --engage", async (t) => {
  using console = fakeConsole();
  using fetch = mockFetch(t);
  await main([...config(fetch), "feed", "--engage"], { path: ":memory:" });
  await assertSnapshot(t, console.calls);
});

Deno.test("duolingo follows", async (t) => {
  using console = fakeConsole();
  using fetch = mockFetch(t);
  await main([...config(fetch), "follows"], { path: ":memory:" });
  await assertSnapshot(t, console.calls);
});

Deno.test("duolingo follows --follow", async (t) => {
  using console = fakeConsole();
  using fetch = mockFetch(t);
  await main([...config(fetch), "follows", "--follow"], { path: ":memory:" });
  await assertSnapshot(t, console.calls);
});

Deno.test("duolingo follows --unfollow", async (t) => {
  using console = fakeConsole();
  using fetch = mockFetch(t);
  await main([...config(fetch), "follows", "--unfollow"], { path: ":memory:" });
  await assertSnapshot(t, console.calls);
});

Deno.test("duolingo league", async (t) => {
  using console = fakeConsole();
  using fetch = mockFetch(t);
  await main([...config(fetch), "league"], { path: ":memory:" });
  await assertSnapshot(t, console.calls);
});

Deno.test("duolingo league --follow", async (t) => {
  using console = fakeConsole();
  using fetch = mockFetch(t);
  await main([...config(fetch), "league", "--follow"], { path: ":memory:" });
  await assertSnapshot(t, console.calls);
});
