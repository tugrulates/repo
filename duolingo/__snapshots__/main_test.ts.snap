export const snapshot = {};

snapshot[`duolingo feed 1`] = `
[
  {
    data: [
      "💪 Mary collected the march badge!",
    ],
    level: "log",
  },
  {
    data: [
      "💪 Ana Lucia Santos collected the march badge!",
    ],
    level: "log",
  },
  {
    data: [
      "💪 Sue collected the march badge!",
    ],
    level: "log",
  },
  {
    data: [
      "🎉 Victor Hugo completed 30 friends quests in a row!",
    ],
    level: "log",
  },
  {
    data: [
      "💪 Victor Hugo collected the march badge!",
    ],
    level: "log",
  },
  {
    data: [
      "🎉 azra completed a 7 day streak!",
    ],
    level: "log",
  },
  {
    data: [
      "💪 Anna collected the march badge!",
    ],
    level: "log",
  },
  {
    data: [
      "🎉 Fiona completed 5 friends quests in a row!",
    ],
    level: "log",
  },
  {
    data: [
      "💪 Sonnenschein collected the march badge!",
    ],
    level: "log",
  },
  {
    data: [
      "🎉 abc completed 5 friends quests in a row!",
    ],
    level: "log",
  },
  {
    data: [
      "🎉 Namık Öztürk completed 5 friends quests in a row!",
    ],
    level: "log",
  },
  {
    data: [
      "🎉 its_aymanYT completed 3 friends quests in a row!",
    ],
    level: "log",
  },
  {
    data: [
      "💪 GULCAN collected the march badge!",
    ],
    level: "log",
  },
  {
    data: [
      "🎉 Pilibep completed 25 friends quests in a row!",
    ],
    level: "log",
  },
  {
    data: [
      "🎉 Pilibep did more than 10 lessons in a day!",
    ],
    level: "log",
  },
  {
    data: [
      "🎉 Laura completed 3 friends quests in a row!",
    ],
    level: "log",
  },
  {
    data: [
      "💪 Kim collected the march badge!",
    ],
    level: "log",
  },
  {
    data: [
      "💪 Robby A collected the march badge!",
    ],
    level: "log",
  },
  {
    data: [
      "🎉 Kat completed 5 friends quests in a row!",
    ],
    level: "log",
  },
  {
    data: [
      "🎉 Jasper did more than 10 lessons in a day!",
    ],
    level: "log",
  },
  {
    data: [
      "🙏 Tuğrul you and kathrynelsarraff completed the friends quest!",
    ],
    level: "log",
  },
  {
    data: [
      "💪 Ilona collected the march badge!",
    ],
    level: "log",
  },
  {
    data: [
      "💪 Ciflorette collected the march badge!",
    ],
    level: "log",
  },
  {
    data: [
      "💪 Therese Saade Karam collected the march badge!",
    ],
    level: "log",
  },
  {
    data: [
      "💪 frankrae collected the march badge!",
    ],
    level: "log",
  },
  {
    data: [
      "💪 I want 500 subscribers collected the march badge!",
    ],
    level: "log",
  },
  {
    data: [
      "💪 Gerda collected the march badge!",
    ],
    level: "log",
  },
  {
    data: [
      "👍 RightPlaceWrongPersonByRM shared a sentence",
    ],
    level: "log",
  },
  {
    data: [
      "👍 RightPlaceWrongPersonByRM shared a sentence",
    ],
    level: "log",
  },
  {
    data: [
      "💪 Thomé collected the march badge!",
    ],
    level: "log",
  },
  {
    data: [
      "💪 Арина collected the march badge!",
    ],
    level: "log",
  },
  {
    data: [
      "💪 zeynep h ates collected the march badge!",
    ],
    level: "log",
  },
  {
    data: [
      "💪 Akipapi collected the march badge!",
    ],
    level: "log",
  },
  {
    data: [
      "🎉 Akipapi completed 20 friends quests in a row!",
    ],
    level: "log",
  },
  {
    data: [
      "🎉 I want 500 subscribers completed 3 friends quests in a row!",
    ],
    level: "log",
  },
  {
    data: [
      "💯 drumwhacker completed a 1500 day streak!",
    ],
    level: "log",
  },
  {
    data: [
      "💪 Юлия collected the march badge!",
    ],
    level: "log",
  },
  {
    data: [
      "💪 Atayvia Roberts collected the march badge!",
    ],
    level: "log",
  },
  {
    data: [
      "💪 Susi collected the march badge!",
    ],
    level: "log",
  },
  {
    data: [
      "💪 Esther collected the march badge!",
    ],
    level: "log",
  },
  {
    data: [
      "💪 Melanie collected the march badge!",
    ],
    level: "log",
  },
  {
    data: [
      "💪 helma auler collected the march badge!",
    ],
    level: "log",
  },
  {
    data: [
      "💪 Sima collected the march badge!",
    ],
    level: "log",
  },
  {
    data: [
      "💪 Ingrid collected the march badge!",
    ],
    level: "log",
  },
  {
    data: [
      "💪 RJracoon collected the march badge!",
    ],
    level: "log",
  },
  {
    data: [
      "🎉 Sheri Poisson ONeill did more than 10 lessons in a day!",
    ],
    level: "log",
  },
]
`;

snapshot[`duolingo feed --engage 1`] = `
[
  {
    data: [
      "💪 Mary collected the march badge!",
    ],
    level: "log",
  },
  {
    data: [
      "💪 Ana Lucia Santos collected the march badge!",
    ],
    level: "log",
  },
  {
    data: [
      "💪 Sue collected the march badge!",
    ],
    level: "log",
  },
  {
    data: [
      "🎉 Victor Hugo completed 30 friends quests in a row!",
    ],
    level: "log",
  },
  {
    data: [
      "💪 Victor Hugo collected the march badge!",
    ],
    level: "log",
  },
  {
    data: [
      "🎉 azra completed a 7 day streak!",
    ],
    level: "log",
  },
  {
    data: [
      "💪 Anna collected the march badge!",
    ],
    level: "log",
  },
]
`;

snapshot[`duolingo follows 1`] = `
[
  {
    data: [
      "👤 Following 384 people.",
    ],
    level: "log",
  },
  {
    data: [
      "👤 Followed by 397 people.",
    ],
    level: "log",
  },
]
`;

snapshot[`duolingo follows --follow 1`] = `
[
  {
    data: [
      "👤 Following 384 people.",
    ],
    level: "log",
  },
  {
    data: [
      "👤 Followed by 397 people.",
    ],
    level: "log",
  },
]
`;

snapshot[`duolingo follows --unfollow 1`] = `
[
  {
    data: [
      "❌ Unfollowed AgaKrzew.",
    ],
    level: "log",
  },
  {
    data: [
      "❌ Unfollowed Crodragon2585.",
    ],
    level: "log",
  },
  {
    data: [
      "❌ Unfollowed Mattias697103.",
    ],
    level: "log",
  },
  {
    data: [
      "❌ Unfollowed Pt9Yex.",
    ],
    level: "log",
  },
  {
    data: [
      "👤 Following 380 people.",
    ],
    level: "log",
  },
  {
    data: [
      "👤 Followed by 397 people.",
    ],
    level: "log",
  },
]
`;

snapshot[`duolingo league 1`] = `
[
  {
    data: [
      "🩷  Pearl League                   
 1. Tuğrul 👀               1799 XP
 2. Crodragon 🍿            1713 XP
 3. Pöt                     1644 XP
 4. Aga Krzew               1201 XP
 5. Mattias                 1103 XP
 6. wotan 😡                 841 XP
 7. Simone 🇮🇹                797 XP
 8. Valko Gada 💩            747 XP
 9. Ira                      574 XP
10. Lucka                    559 XP
11. Büsra Dölek          👤  548 XP
12. Marcelo Costa Filho  👤  528 XP
13. Steen                    523 XP
14. G                        522 XP
15. Luke 😡                  434 XP
16. Steph                    352 XP
17. Luna                     323 XP
18. Jirka                    288 XP
19. Minttu                   251 XP
20. Andrea 😡                213 XP",
    ],
    level: "log",
  },
]
`;

snapshot[`duolingo league --follow 1`] = `
[
  {
    data: [
      "🩷  Pearl League                   
 1. Tuğrul 👀               1799 XP
 2. Crodragon 🍿            1713 XP
 3. Pöt                     1644 XP
 4. Aga Krzew               1201 XP
 5. Mattias                 1103 XP
 6. wotan 😡                 841 XP
 7. Simone 🇮🇹                797 XP
 8. Valko Gada 💩            747 XP
 9. Ira                      574 XP
10. Lucka                    559 XP
11. Büsra Dölek          👤  548 XP
12. Marcelo Costa Filho  👤  528 XP
13. Steen                    523 XP
14. G                        522 XP
15. Luke 😡                  434 XP
16. Steph                    352 XP
17. Luna                     323 XP
18. Jirka                    288 XP
19. Minttu                   251 XP
20. Andrea 😡                213 XP",
    ],
    level: "log",
  },
]
`;
